File: app/src/main/java/com/wang/javamvp/base/rx/RxHelper.java
package com.wang.javamvp.base.rx;

import com.wang.javamvp.base.bean.BaseBean;
import com.wang.javamvp.base.rx.errorbean.ApiException;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

/**
 * @author Mis Wang
 * @date 2018/5/14  14:02
 * @fuction Observable转换器
 */
public class RxHelper {

    public static <T> ObservableTransformer<BaseBean<T>, T> handleResult() {
        return new ObservableTransformer<BaseBean<T>, T>() {
            @Override
            public ObservableSource<T> apply(Observable<BaseBean<T>> upstream) {
                return upstream.flatMap(
                        new Function<BaseBean<T>, ObservableSource<T>>() {
                            @Override
                            public ObservableSource<T> apply(BaseBean<T> tBaseBean) throws Exception {
                                if (!tBaseBean.isError()) {
                                    return createObservable(tBaseBean.getResults());
                                } else {
                                    return Observable.error(new ApiException("-1", "错误数据"));
                                }
                            }
                        }).subscribeOn(Schedulers.io())
                        .unsubscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread());
            }
        };
    }


    private static <T> Observable<T> createObservable(final T t) {
        return Observable.create(new ObservableOnSubscribe<T>() {
            @Override
            public void subscribe(ObservableEmitter<T> emitter) throws Exception {
                try {
                    emitter.onNext(t);
                    emitter.onComplete();
                } catch (Exception e) {
                    emitter.onError(e);
                }
            }
        });
    }
}

File: app/src/main/java/com/wang/kotlinmvp/base/rx/RxHelper.kt
package com.wang.kotlinmvp.base.rx

import com.wang.kotlinmvp.base.bean.KotlinBean
import com.wang.kotlinmvp.base.rx.errorbean.ApiException
import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers


/**
 * @author Mis Wang
 * @date  2018/5/11  15:14
 * @fuction Observable转换器
 */
object RxHelper {
    fun <T> handleResult(): ObservableTransformer<KotlinBean<T>, T> {
        return ObservableTransformer { upstream ->
            upstream.flatMap { result ->
                //                if (ConstantUtil.SUCCESS.equals(result.status.code)) {
//                    createObservable(result.content);
//                } else {
//                    Observable.error(ApiException(result.status.code, result.status.message));
//                }
                if (!result.error) {
                    createObservable(result.results);
                } else {
                    Observable.error(ApiException("-1", "错误数据"));
                }
            }.subscribeOn(Schedulers.io())
                    .unsubscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
        }
    }

    /**
     * 创建被观察者
     */
    private fun <T> createObservable(t: T): Observable<T> {
        return Observable.create({ emitter ->
            try {
                emitter.onNext(t)
                emitter.onComplete()
            } catch (e: Exception) {
                e.printStackTrace()
                emitter.onError(e)
            }
        })
    }
}

